modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala

```diff
@@ -153,28 +220,150 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
+    nameTransformation: Option[c.Expr[String => String]],
+    useDefaults: c.Expr[Boolean],
+    discriminator: c.Expr[Discriminator]
   ): c.Expr[Decoder[T]] = {
     val tpe = weakTypeOf[T]
 
-    def transformName(name: String): Tree = nameTransformation.fold[Tree](q"$name")(f => q"$f($name)")
+    val subclasses = tpe.typeSymbol.asClass.knownDirectSubclasses
+    if (subclasses.isEmpty) {
+      materializeDecoderCaseClassImpl[T](nameTransformation, useDefaults)
+    } else {
+      materializeDecoderTraitImpl[T](
+        nameTransformation,
+        subclasses,
+        discriminator
+      )
+    }
+  }
+
+  private[this] def materializeDecoderTraitImpl[T: c.WeakTypeTag](
+    nameTransformation: Option[c.Expr[String => String]],
+    subclasses: Set[Symbol],
+    discriminator: c.Expr[Discriminator]
+  ): c.Expr[Decoder[T]] = {
+    val tpe = weakTypeOf[T]
+
+    expandDiscriminator(discriminator.tree) match {
+      case _root_.io.circe.derivation.Discriminator.Embedded(fieldName) =>
+        val instanceDefs: List[Tree] = subclasses.map { s =>
+          val value =
+            Literal(Constant(s.asClass.name.decodedName.toString.toLowerCase()))
+
+          cq""" $value => _root_.io.circe.Decoder[${s.asType}]"""
+        }.toList
+
+        c.Expr[Decoder[T]](q"""
+            for {
+                visitorType <- _root_.io.circe.Decoder[String].prepare(_.downField($fieldName))
+                value <- visitorType match {
+                  case ..$instanceDefs
+                }
+              } yield value
+
+          """)
+      case _root_.io.circe.derivation.Discriminator.TypeDiscriminator =>
+        val instanceDefs: List[Tree] = subclasses.map { s =>
+          val value =
+            Literal(Constant(s.asClass.name.decodedName.toString.toLowerCase()))
+
+          cq""" $value => c.get(keys.head)(_root_.io.circe.Decoder[${s.asType}]).asInstanceOf[ _root_.io.circe.Decoder.Result[$tpe]]"""
+        }.toList
+
+        val result = q"""
+           new _root_.io.circe.Decoder[$tpe] {
+                 override def apply(c: _root_.io.circe.HCursor): _root_.io.circe.Decoder.Result[$tpe] = {
+                   val keys=c.keys.map(_.toList).getOrElse(Nil)
+                   if(keys.isEmpty){
+                     Left(_root_.io.circe.DecodingFailure("Missing type field discriminator for trait field", Nil))
+                   } else {
+                     keys.head match {
+                       case ..$instanceDefs
+                     }
+                   }
+                 }
+               }
+          """
+        c.Expr[Decoder[T]](result)
+    }
+
+  }
+
+  private[this] def materializeDecoderCaseClassImpl[T: c.WeakTypeTag](
+    nameTransformation: Option[c.Expr[String => String]],
+    useDefaults: c.Expr[Boolean]
+  ): c.Expr[Decoder[T]] = {
+    val tpe = weakTypeOf[T]
+
+    // Valid only in macro!!
+    val globalUseDefaults: Boolean = extractUseDefaults(useDefaults.tree)
```

Comment:
Not sure about the name here. In what sense is it global?

Reply:
I'll rename in *useDefaults*